### ISSUE TITLE
maxTimeMS not working for Find method

### ISSUE THREAD

**Issue by olawale-fishpond** [mongodb/mongo-php-library#346]
Hi,
Is this a contradiction or the code implementation is different?
See mongo-php-library/src/Operation/Find.php:263

> return new Query($this->filter, $options);

and  https://github.com/mongodb/mongo-php-driver/issues/235#issuecomment-187358228
> $query = new MongoDB\Driver\Query(
    ['foo' => 'bar'],
    ['modifiers' => ['$maxTimeMS' => 1000]]
);

That is, the way maxTimeMS was passed to the Query is different from what the comment recommended.
Excerpt from https://github.com/mongodb/mongo-php-driver/issues/235#issuecomment-187358228

> Specifying maxTimeMS is different for commands and queries (see: server docs). For queries, maxTimeMS is set through the $maxTimeMS modifier, which would be set through the Query object's modifiers
It seems the mongo-php-library/src/Operation/Find.php:263 treated the 'maxTimeMS' in a form
that the Command object will be returned.

> See mongo-php-library/src/Operation/Count.php:186 i.e. return new Command($cmd);

Could the  'maxTimeMS' option at Find.php:247 be  stored as one of the values for $options['modifiers']? At the moment,  specified timeout via 'maxTimeMS' option is not working.

Environment
mongo 2.6.12

**Comment by jmikola** [comment]
> Could the 'maxTimeMS' option at Find.php:247 be stored as one of the values for $options['modifiers']?

Specifying `maxTimeMS` via the top-level option should have the same effect as specifying `$maxTimeMS` in the `modifiers` option; however, the latter is deprecated. The relevant driver ticket where `modifiers` were deprecated is [PHPC-783](https://jira.mongodb.org/browse/PHPC-783) (released in 1.2.0) and underlying libmongoc ticket [CDRIVER-1522](https://jira.mongodb.org/browse/CDRIVER-1522). The driver reads both options and determines how to issue the query based on whether the server supports the [`find` command](https://docs.mongodb.com/manual/reference/command/find/). In the driver documentation, [`MongoDB\Driver\Query::__construct()`](http://php.net/manual/en/mongodb-driver-query.construct.php) mentions that `modifiers` are deprecated and that top-level options will take precedence. I realize we do not include the same notice in the [`MongoDB\Collection::find()`](https://docs.mongodb.com/php-library/master/reference/method/MongoDBCollection-find/) documentation, so I'll update that soon to say the same.

> At the moment, specified timeout via 'maxTimeMS' option is not working.

I attempted to reproduce this using MongoDB 2.6.12, PHP driver 1.2.7, and PHP library 1.1.2 with the following script:

```php
<?php

require 'vendor/autoload.php';

$collection = (new MongoDB\Client)->test->foo;
$collection->drop();

$documents = [];

for ($i = 0; $i < 10000; $i++) {
    $documents[] = ['x' => $i];
}
$result = $collection->insertMany($documents);
var_dump($result->getInsertedCount());

$cursor = $collection->find(['x'=>['$gte'=>5000]], ['maxTimeMS' => 1]);
var_dump(iterator_count($cursor));
```

The output was as expected:

```
int(10000)
PHP Fatal error:  Uncaught MongoDB\Driver\Exception\ExecutionTimeoutException: operation exceeded time limit in /home/jmikola/workspace/mongodb/phplib/src/Operation/Find.php:219
```

What evidence do you have that the `maxTimeMS` option is not working?

**Comment by jmikola** [comment]
> I realize we do not include the same notice in the `MongoDB\Collection::find()` documentation, so I'll update that soon to say the same.

I just merged https://github.com/mongodb/mongo-php-library/pull/347, which should be published later today. I also opened [PHPLIB-264](https://jira.mongodb.org/browse/PHPLIB-264) to address top-level options that are supported on `MongoDB\Driver\Query` but missing in the library. As-is, users still need to use `modifiers` in the library to supply some options to the driver (e.g. `hint`, `explain`).

**Comment by olawale-fishpond** [comment]
> What evidence do you have that the maxTimeMS option is not working?

Environment:  PHP 7.0.13-0ubuntu0.16.04.1 (cli) ( NTS ), MongoDB 2.6.12, PHP ext 1.1.5, and PHP library 1.1.0

Using your sample code above, there was no ExecutionTimeoutException but using below did.
$cursor = $collection->find(['x'=>['$gte'=>5000]], ['modifiers' => ['$maxTimeMS' => 1]]);

**Comment by jmikola** [comment]
> PHP ext 1.1.5, and PHP library 1.1.0

PHP library 1.1.0 requires [`ext-mongodb` 1.2.0 or later](https://github.com/mongodb/mongo-php-library/blob/1.1.0/composer.json#L14). Did you update the library through Composer or did you manually install it? If the former, I expect Composer would have reported the unsatisfied extension dependency.

**Comment by olawale-fishpond** [comment]
Composer was used with requirement to ignore such i.e. php composer.phar --ignore-platform-reqs --no-dev install.
Alright, so it's an issue of differences between the ext-mongodb and the PHP library?

**Comment by jmikola** [comment]
> so it's an issue of differences between the ext-mongodb and the PHP library?

Correct. I alluded to [PHPC-783](https://jira.mongodb.org/browse/PHPC-783) (released in 1.2.0) earlier in https://github.com/mongodb/mongo-php-library/issues/346#issuecomment-287121445. That is the driver ticket where we deprecated the `modifiers` option and added support for the various top-level options that correspond to old modifiers. Upgrading the extension should resolve this for you.

If you're going to ignore platform requirements with Composer, you should ensure that extensions are updated when needed (both for this library and potentially other packages).
